django/core/cache/backends/redis.py

```diff
@@ -0,0 +1,242 @@
+        index = self._get_connection_pool_index(write)
+
+        params = self._parse_url(self._servers[index])
+        params.update(self._client_kwargs)
+
+        if self._pools[index] is None:
+            if self._servers[index].startswith("unix:"):
+                self._pools[index] = self._pool_class(
+                    connection_class=redis.UnixDomainSocketConnection,
+                    path=self._servers[index][5:],
+                    **self._client_kwargs
+                )
+            else:
+                self._pools[index] = self._pool_class(**params)
+
+        return self._pools[index]
+
+    def get_client(self, key="", write=False):
+        pool = self._get_connection_pool(write)
+        return self._client(connection_pool=pool)
+
+    def get(self, key):
+        client = self.get_client(key)
+        value = client.get(key)
+        # If key does not exists, return None
+        if value is None:
+            return value
+        return self._serializer.loads(value)
+
+    def add(self, key, value, timeout):
+        client = self.get_client(key)
+        return bool(client.set(key, self._serializer.dumps(value), ex=timeout, nx=True))
+
+    def set(self, key, value, timeout):
+        client = self.get_client(key, write=True)
+        client.set(key, self._serializer.dumps(value), ex=timeout)
+
+    def touch(self, key, timeout):
+        client = self.get_client(key, write=True)
+        if timeout is None:
+            return client.persist(key)
+        else:
+            return client.expire(key, timeout)
+
+    def delete(self, key):
+        client = self.get_client(key, write=True)
+        return bool(client.delete(key))
+
+    def clear(self):
+        client = self.get_client(None, write=True)
+        return client.flushall()
+
+    def get_many(self, keys):
+        client = self.get_client(None)
+        ret = client.mget(keys)
+        return {
+            k: self._serializer.loads(v) for k, v in zip(keys, ret) if v is not None
+        }
+
+    def set_many(self, data, timeout):
+        client = self.get_client(None)
+        serialized_data = {k: self._serializer.dumps(v) for k, v in data.items()}
+
+        client.mset(serialized_data)
+
+        for k, _ in data.items():
+            # Setting timeout for each as redis-py does not support timeout with mset
+            if timeout is None:
+                client.persist(k)
+            else:
+                client.expire(k, timeout)
+
+
+class RedisCache(BaseCache):
+    def __init__(self, server, params):
+        super().__init__(params)
+        if isinstance(server, str):
+            self._servers = re.split("[;,]", server)
+        else:
+            self._servers = server
+
+        self._options = params.get("OPTIONS") or {}
+        self._client_class = self._options.get("CLIENT", RedisCacheClient)
+
+        if isinstance(self._client_class, str):
+            self._client_class = import_string(self._client_class)
```

Comment:
I kept this as in my GSoC proposal, I wanted to implement two client classes. For multiple cache server, one would follow a replication based approach and the other would be sharding based. However, this can be added later on when (if) the other client might be implemented. So, should I remove this for now?

Reply:
I understand. Although I'm wondering if this is something that can just be configured by passing something in `OPTIONS` that is passed to the client class such that you don't need two separate client classes?

(Failing that, a separate backend, e.g. `RedisShardedCache`, might make more sense? I know we are implementing the client class here, but for all other backends we are using an existing client class from a third-party package - we are never exposing knowledge of the client class to the end user.)

Either way, if that sharded implementation is coming later (whether in another commit or PR), let's not add this client class loading stuff now. It doesn't add any benefit to this initial implementation other than making it more complex to review.